src/rust/engine/fs/store/src/remote.rs

```diff
@@ -14,33 +13,37 @@ use hashing::Digest;
 use log::Level;
 use protos::gen::build::bazel::remote::execution::v2 as remexec;
 use remexec::ServerCapabilities;
-use tokio::io::{AsyncSeekExt, AsyncWrite};
+use tokio::io::{AsyncRead, AsyncSeekExt, AsyncWrite};
 use workunit_store::{in_workunit, ObservationMetric};
 
-use crate::StoreError;
-
 mod reapi;
 #[cfg(test)]
 mod reapi_tests;
 
-pub type ByteSource = Arc<(dyn Fn(Range<usize>) -> Bytes + Send + Sync + 'static)>;
+pub type StoreSource = Box<dyn AsyncRead + Send + Sync + Unpin + 'static>;
 
 #[async_trait]
 pub trait ByteStoreProvider: Sync + Send + 'static {
-  async fn store_bytes(&self, digest: Digest, bytes: ByteSource) -> Result<(), String>;
+  /// Store the bytes readable from `source` into the remote store
+  async fn store(&self, digest: Digest, source: StoreSource) -> Result<(), String>;
+
+  /// Store the bytes in `bytes` into the remote store, as an optimisation of `store` when the bytes
+  /// are already in memory
+  async fn store_bytes(&self, digest: Digest, bytes: Bytes) -> Result<(), String>;
```

Comment:
I think there's a fair argument that there's no need for `store_bytes` and everything can go through `store`, using `Cursor::new(bytes)` to create an appropriate `AsyncRead` from `Bytes`.

My thinking is that passing `Bytes` in directly saves some memory copies for the batch case, where that object can be splatted into  `store_bytes_batch` and its batch upload request directly, without copying or slicing or anything (whereas using `source` would require reading it into a separate `Bytes`).

Maybe that optimisation is irrelevant when this code does network IO anyway, and it'd be better to just have this trait be `store`, `load` and `list_missing_digests`.

Thoughts?

Reply:
AFAICT, you're right that it is one fewer copy currently to pass in `Bytes`. As mentioned in the comment on `store_lmdb_blob_remote` though, we used to write in a streaming fashion while blocking a `spawn_blocking` task (with `block_on`): that meant that we were copying directly from a MMAP into another buffer.

It's possible that we could re-introduce that optimization at some point, which would make the streaming `store` API superior again. But at the same time, these are fairly small blobs, so the benefits of streaming are definitely reduced.